src/org/sosy_lab/common/collect/OrderStatisticSet.java

```diff
@@ -0,0 +1,150 @@
+   * 0, the largest element in the set has rank == size - 1.
+   *
+   * @param pIndex the rank of the element to return
+   * @return the element of this set with the given rank
+   * @throws IndexOutOfBoundsException if the given rank is out of the range of this set (i.e.,
+   *     pRank &lt; 0 || pRank &gt;= size)
+   */
+  E getByRank(int pIndex);
+
+  /**
+   * Remove the element of this set with the given rank and return it.
+   *
+   * <p>The lowest element in the set has rank == 0, the largest element in the set has rank == size
+   * - 1.
+   *
+   * @param pIndex the rank of the element to remove
+   * @return the removed element
+   * @throws IndexOutOfBoundsException if the given rank is out of the range of this set (i.e.,
+   *     pRank &lt; 0 || pRank &gt;= size)
+   * @see #getByRank(int)
+   */
+  @CanIgnoreReturnValue
+  E removeByRank(int pIndex);
+
+  /**
+   * Return the rank of the given element in this set. Returns -1 if the element does not exist in
+   * the set.
+   *
+   * <p>The lowest element in the set has rank == 0, the largest element in the set has rank == size
+   * - 1.
+   *
+   * @param pObj the element to return the rank for
+   * @return the rank of the given element in the set, or -1 if the element is not in the set
+   * @throws NullPointerException if the given element is <code>null</code>
+   */
+  int rankOf(E pObj);
+
+  @Override
+  OrderStatisticSet<E> descendingSet();
+
+  @Override
+  OrderStatisticSet<E> subSet(
+      E pFromElement, boolean fromInclusive, E pToElement, boolean toInclusive);
+
+  @Override
+  OrderStatisticSet<E> headSet(E pToElement, boolean inclusive);
+
+  @Override
+  OrderStatisticSet<E> tailSet(E pFromElement, boolean inclusive);
+
+  @Override
+  OrderStatisticSet<E> subSet(E pFromElement, E pToElement);
+
+  @Override
+  OrderStatisticSet<E> headSet(E pToElement);
+
+  @Override
+  OrderStatisticSet<E> tailSet(E pFromElement);
+
+  /**
+   * Creates a new empty OrderStatisticSet using natural ordering. The returned map guarantees
+   * performance only in O(n) for the operations specific to the OrderStatisticSet interface.
+   */
+  static <E> OrderStatisticSet<E> create() {
+    return NaiveOrderStatisticSet.createSet();
+  }
+
+  /**
+   * Creates a new empty OrderStatisticSet using the given comparator. The returned map guarantees
+   * performance only in O(n) for the operations specific to the OrderStatisticSet interface.
+   */
+  static <E> OrderStatisticSet<E> create(Comparator<? super E> pComparator) {
+    return NaiveOrderStatisticSet.createSet(pComparator);
+  }
+
+  /**
+   * Creates a new OrderStatisticSet containing the same elements as the given collection, using
+   * natural ordering. The returned map guarantees performance only in O(n) for the operations
+   * specific to the OrderStatisticSet interface.
+   */
+  static <E> OrderStatisticSet<E> createWithNaturalOrder(Collection<E> pCollection) {
```

Comment:
We could accept an `Iterable` here. By using `Iterables.addAll`, we do not need more code and get the same performance benefit as before if a `TreeMap` is passed.